### ISSUE TITLE
Logs not being forwarded

### ISSUE THREAD

**Issue by arelstone** [newrelic/newrelic-react-native-agent#195]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description
I have been awaiting log forwarding  in a very long time and as soon as i noticed it in the documentation I did a tiny experiment... Adding some log statements

I have not been able to see any logs from my react native app.

When I do a NRQL `FROM Log SELECT message WHERE entity.name LIKE '` the service does not show up in the list of suggestions.
No logs show up if in the NR interface select `Mobile` -> Project -> `Logs` and change the time to be within the last 24 hours.

## Your Environment

<details>
info Fetching system and libraries information...
System:
  OS: macOS 14.6.1
  CPU: (8) arm64 Apple M1 Pro
  Memory: 1.61 GB / 32.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.10.0
    path: ~/.nvm/versions/node/v20.10.0/bin/node
  Yarn:
    version: 3.8.2
    path: /opt/homebrew/bin/yarn
  npm:
    version: 10.2.3
    path: ~/.nvm/versions/node/v20.10.0/bin/npm
  Watchman:
    version: 2023.12.04.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /Users/...../.rbenv/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.4
      - iOS 17.4
      - macOS 14.4
      - tvOS 17.4
      - visionOS 1.1
      - watchOS 10.4
  Android SDK:
    API Levels:
      - "30"
      - "31"
      - "33"
      - "33"
      - "34"
    Build Tools:
      - 29.0.2
      - 30.0.2
      - 30.0.3
      - 31.0.0
      - 33.0.0
      - 33.0.1
      - 34.0.0
    System Images:
      - android-33 | Google APIs ARM 64 v8a
    Android NDK: Not Found
IDEs:
  Android Studio: 2024.1 AI-241.18034.62.2411.12169540
  Xcode:
    version: 15.3/15E204a
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.9
    path: /usr/bin/javac
  Ruby:
    version: 3.2.1
    path: /Users/...../.rbenv/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.3.1
    wanted: 18.3.1
  react-native:
    installed: 0.74.3
    wanted: 0.74.3
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false
</details>

## Additional context
**Versions:**

package.json
```
"react-native": "0.74.3",
"newrelic-react-native-agent": "1.4.3",
```

Podfile.lock:
```lock
- newrelic-react-native-agent (1.4.3):
    - NewRelicAgent (~> 7.5.1)
    - React
```

StartAgent:
```js
NewRelic.startAgent(Platform.OS === 'ios' ? Config.IOS_NEWRELIC_TOKEN : Config.ANDROID_NEWRELIC_TOKEN, {
  analyticsEventEnabled: true,
  nativeCrashReportingEnabled: true,
  crashReportingEnabled: true,
  interactionTracingEnabled: true,
  networkRequestEnabled: true,
  networkErrorRequestEnabled: true,
  httpRequestBodyCaptureEnabled: true,
  httpResponseBodyCaptureEnabled: true,
  loggingEnabled: true,
  webViewInstrumentation: true,
  offlineStorageEnabled: true,
  newEventSystemEnabled: true,
  distributedTracingEnabled: true,
});
NewRelic.setAttribute('environment', Config.ENVIRONMENT);
NewRelic.setJSAppVersion(version);
```

Example of log send
```js
 NewRelic.logVerbose(`Request: "${config.method?.toUpperCase()}" ${config.url}`);
```

Please note that `Config.IOS_NEWRELIC_TOKEN`, `Config.ANDROID_NEWRELIC_TOKEN` and  `Config.ENVIRONMENT` IS working as expected

**Comment by bdotsamir** [comment]
also experiencing this bug. I picked newrelic because of its logging feature! this is critical to me

**Comment by bdotsamir** [comment]
can also confirm that this is happening in both development builds (using expo dev client) and in production builds (without the dev client)

**Comment by bdotsamir** [comment]
Update: I've written up a small and hacky workaround. I have overridden my most-used console functions (console.log, .warn, .error, etc), and now post messages to the NewRelic logging api.

```ts
import { inspect } from "util";

const defaultConsoleLog = console.log;
const defaultConsoleWarn = console.warn;
const defaultConsoleError = console.error;
const defaultConsoleInfo = console.info;
const defaultConsoleDebug = console.debug;

export async function sendToNewRelic(level: string, message: string) {
  return new Promise((resolve, reject) => {
    fetch(`https://log-api.newrelic.com/log/v1?Api-Key=${process.env.EXPO_PUBLIC_NEW_RELIC_API_KEY}`, {
      method: "POST",
      body: JSON.stringify({
        level,
        message
      })
    }).then(res => {
      if (res.ok) {
        resolve(res);
      } else {
        defaultConsoleError("Error sending log to new relic: ", res);
        reject(res);
      }
    }).catch(err => {
      defaultConsoleError("Error sending log to new relic: ", err);
      reject(err);
    });
  });
}

/* eslint-disable @typescript-eslint/no-explicit-any */
function cleanArgs(args: any[]) {
  return args.map(arg => {
    if (typeof arg === "object") {
      return inspect(arg, { depth: 1 });
    } else {
      return arg;
    }
  });
}

console.log = (...args: any[]) => {
  defaultConsoleLog(...args);

  const cleanedArgs = cleanArgs(args).join(" ");
  sendToNewRelic("info", inspect(cleanedArgs, { depth: 1 }));
};

console.warn = (...args) => {
  defaultConsoleWarn(...args);

  const cleanedArgs = cleanArgs(args).join(" ");
  sendToNewRelic("warn", inspect(cleanedArgs, { depth: 1 }));
};

console.error = (...args) => {
  defaultConsoleError(...args);

  const cleanedArgs = cleanArgs(args).join(" ");
  sendToNewRelic("error", inspect(cleanedArgs, { depth: 1 }));
};

console.info = (...args) => {
  defaultConsoleInfo(...args);

  const cleanedArgs = cleanArgs(args).join(" ");
  sendToNewRelic("info", inspect(cleanedArgs, { depth: 1 }));
};

console.debug = (...args) => {
  defaultConsoleDebug(...args);

  const cleanedArgs = cleanArgs(args).join(" ");
  sendToNewRelic("debug", inspect(cleanedArgs, { depth: 1 }));
};
/* eslint-enable @typescript-eslint/no-explicit-any */
```

This is not ideal. I would rather not have to inject my own logic into global functions that are very often used. This method also loses quite a lot of useful metadata that I believe NewRelic would otherwise capture, but it's better than nothing.

I also tried my hand at debugging this library to see *where* along the chain the messages were getting lost, but the debugger took me all the way to the bridge between JS <-> native code and I still hadn't found the issue without having to plunge into native-land. Unfortunately, though I am skilled at debugging, I am not skilled enough to beat my debugger into following native code in addition to JS-land code, and so the trail went cold.

**Comment by arelstone** [comment]
Thanks @bdotsamir. I agree with you it does seem like a lot of useful metadata is getting lost :(

Honestly it feels like the newrelic team doesn't care about the companies that uses react native. Sadly :( My hands are tied to new relic because of management decisions and history.

**Comment by ndesai-newrelic** [comment]
@arelstone @bdotsamir did you enable log reporting on application settings page?

**Comment by arelstone** [comment]
> @arelstone @bdotsamir did you enable log reporting on application settings page?

Yea, on my side log collecting is enabled @ndesai-newrelic

**Comment by bdotsamir** [comment]
> @arelstone @bdotsamir did you enable log reporting on application settings page?

Hello, thank you for following up. My team has actually moved off of New Relic in favor of Data Dog because of an unrelated issue with android clients crashing on exit. #183

**Comment by ndesai-newrelic** [comment]
@bdotsamir the issue is fixed in 1.4.4.

# Regarding the Recent Application Crash Issue

We sincerely apologize for the inconvenience caused by the recent application crashes when transitioning to the background. We understand your frustration, and we'd like to provide a detailed explanation of what occurred and the steps we're taking to prevent similar issues in the future.

## Root Cause

The issue arose due to a complex interaction between our Android and NDK agent releases:

1. We use a "+" convention for version numbering of both Android and NDK agents, which automatically pulls the latest versions during builds.
2. The Android and NDK agents were released in quick succession.
3. Our testing did not catch the scenario where a newer NDK agent was paired with an older Android agent.

This led to a method not found error, initially misdiagnosed as an NDK-side issue. Our attempt to resolve it by reverting NDK changes in an NDK agent release did not address the core problem: a version mismatch between the Android and NDK agents.

## Our Response and Future Prevention

To prevent similar issues in the future, we are implementing the following measures:

1. More comprehensive testing scenarios on version compatibility between components.
2. Modifying our release process to ensure thorough validation of all agent combinations.
3. Reconsidering our use of the "+" convention for version numbers to have more predictable builds.

## Moving Forward

We value your trust and are committed to delivering a stable and reliable React Native agent. We will conduct more thorough testing for future releases to prevent such issues from recurring.

If you need any additional information or clarification regarding this issue, please don't hesitate to reach out. We're here to address any concerns you may have.

Thank you for your patience and understanding as we work to improve our product and processes.

**Comment by bdotsamir** [comment]
@ndesai-newrelic Thank you for the post-mortem. Glad to see that the team is taking things like this seriously. Should the need arise for us to switch off of Data Dog in the future, we'll consider New Relic once again.
